src/main/java/io/cryostat/recordings/RecordingMetadataManager.java

```diff
@@ -177,29 +208,75 @@ public Map<String, String> parseRecordingLabels(String labels) throws IllegalArg
         }
     }
 
-    private Path getMetadataPath(String targetId, String recordingName) {
-        String filename = String.format("%s%s", targetId, recordingName);
+    public void notifyRecordingMetadataUpdated(
+            String targetId, String recordingName, Metadata metadata) {
+        notificationFactory
+                .createBuilder()
+                .metaCategory(RecordingMetadataManager.NOTIFICATION_CATEGORY)
+                .metaType(HttpMimeType.JSON)
+                .message(
+                        Map.of(
+                                "recordingName",
+                                recordingName,
+                                "target",
+                                targetId,
+                                "metadata",
+                                metadata))
+                .build()
+                .send();
+    }
+
+    private Path getMetadataPath(Integer jvmId, String recordingName) {
+        String filename = String.format("%s%s", jvmId, recordingName);
         return recordingMetadataDir.resolve(
                 base32.encodeAsString(filename.getBytes(StandardCharsets.UTF_8)) + ".json");
     }
 
+    private Integer getJvmId(String targetId) throws IOException {
+        Integer jvmId =
+                this.jvmIdMap.computeIfAbsent(
+                        targetId,
+                        k -> {
+                            if (targetId.equals(RecordingArchiveHelper.UNLABELLED)) {
+                                return UNLABELLED_ARCHIVES_ID;
+                            }
+
+                            try {
+                                Credentials credentials =
+                                        credentialsManager.getCredentials(targetId);
+                                return this.targetConnectionManager.executeConnectedTask(
+                                        new ConnectionDescriptor(targetId, credentials),
+                                        connection -> {
+                                            return connection.getJvmId();
+                                        });
+                            } catch (Exception e) {
+                                logger.error(e);
+                                return 0;
+                            }
+                        });
+        if (jvmId == 0) {
```

Comment:
Ah, I see here's the handling for the `ID == 0` case I talked about in the `-core` review for this JVM ID thing. It's valid and feasible for some JVM ID `hashCode` to actually genuinely be 0, or any int value, so this isn't quite safe.

What happens if I have a `targetId` pointing to a valid JVM and I do something (ex. set metadata) that calls this method and populates the `jvmIdMap` for this `targetId`, then that JVM dies and is restarted by the container platform some time later, and then I try to ex. set metadata again using the same `targetId`? It's referencing the same logical JVM, but not *actually* the same JVM instance since it has been restarted and therefore at least one of the MBean attributes will have changed (start time). Since I have this in-memory cache I will reuse the same storage location for these two JVMs. But if the Cryostat instance is then restarted the cache will be empty, and the next time I try to set metadata using this same `targetId` (still referencing the 2nd instance of the restarted target), it will re-populate this cache and come up with a different `jvmId` and therefore a different storage location.

tl;dr the cache lifecycle and the target lifecycle are different, so staleness is tough and needs more consideration here

Reply:
Good point. I'll keep thinking about this.

Reply:
>  But if the Cryostat instance is then restarted the cache will be empty, and the next time I try to set metadata using this same targetId (still referencing the 2nd instance of the restarted target), it will re-populate this cache and come up with a different jvmId and therefore a different storage location.

The `hashCode` [docs](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Object.html#hashCode()) say that different executions of the same application will compute a different integer for the same object. Does that mean that if the Cryostat instance is restarted, and the `jvmId` cache is empty, Cryostat will compute a different `hashCode` for all targets, even for targets that have been running since before the Cryostat instance was restarted?

Reply:
Here's the source for [`HashCodeBuilder`](https://github.com/apache/commons-lang/blob/master/src/main/java/org/apache/commons/lang3/builder/HashCodeBuilder.java). It does indeed end up calling through to the maybe-inconsistent JDK `hashCode()` for general `Object`s, but specifically for primitives and primitive arrays (like `byte[]`) it will be consistent.

Reply:
In case 4 it can become quite expensive to periodically poll, depending on how many discoverable targets there are. Once Cryostat does finish restarting and its platform discovery comes back online, it will "re-discover" any targets around it that were there first, and it will continue to discover any that come online afterward as well. So I think this mostly ends up naturally decomposing into case 1 and case 2. Case 1 and 3 seem simple enough. Case 2 is the hardest one here since you need to notice that a new target coming online is a "replacement" for a previously known target that we had associated with some stored labels.

Reply:
I made an itest in `RecordingMeatdataIT` to test case 2 and was wondering how to generate recording data on an external target?

The test is supposed to create an active recording on a `vertx-fib-demo` container on port 9093, wait for 5 seconds, then archive the recording, but `GET`ting the archived recording always returns `404`. If I try to create a snapshot after waiting 5 seconds, the snapshot also returns `202`, which means there's likely no data generated.

Reply:
If `GET`ing the recording gives a `404` then I think something is wrong, even if the recording is empty it should still be there and give a `200`/`202` rather than a `404`. If you `GET /api/v1/targets/:targetId/recordings` to list the recordings that exist for your external target in test, do you get back an empty list response?

Getting a `202` on snapshot creation when there are no preexisting active recording sources seems to make sense, so that part checks out I think.

Reply:
Hmm, the test case passes if I only use `service:jmx:rmi:///jndi/rmi://%s:9093/jmxrmi` as the target ID, and fails if I also use `Podman.POD_NAME + ":9093"` interchangeably as the target ID. I'll need to think more carefully about how I'm mapping the targets to the metadata.

Reply:
I tried to make the `recordingMetadataManager` implement `Consumer<TargetDiscoveryEvent>` in order to handle stale metadata in `@Override public synchronized void accept(TargetDiscoveryEvent tde)`. It looks like the code inside `accept()` doesn't get invoked based on the itest server logs, though. What am I missing?

Reply:
It looks like you'll need to inject a `PlatformClient` instance into the `RecordingMetadataManager`, and do a `platformClient.addTargetDiscoveryListener(this)`.

Since I'm reading this anyway, now that the Manager also a Verticle you can/should replace the `load()` method with a `start(Promise<void> future)`, so that Vertx performs the initialization on a worker thread rather than `Cryostat.java` doing it on the main application startup thread.